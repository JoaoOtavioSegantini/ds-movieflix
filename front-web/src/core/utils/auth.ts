import jwtDecode from 'jwt-decode'

import history from './history'

export type LoginResponse = {
  access_token: string
  expires_in: number
  scope: string
  token_type: string
  userName: string
  userId: number
}

export type Authority = 'VISITOR' | 'MEMBER' | 'ROLE_ADMIN'

export type AccessToken = {
  exp: number
  user_name: string
  authorities: Authority[]
}

export const saveSessionData = (loginResponse: LoginResponse) => {
  localStorage.setItem('authData', JSON.stringify(loginResponse))
}

export const getSessionData = () => {
  const sessionData = localStorage.getItem('authData') ?? '{}'
  const dataParse = JSON.parse(sessionData)

  return dataParse as LoginResponse
}

export const getAccessTokenDecoded = () => {
  const sessionData = getSessionData()
  try {
    const tokenDecoded = jwtDecode(sessionData.access_token)
    return tokenDecoded as AccessToken
  } catch (err) {
    return {} as AccessToken
  }
}

export const isTokenValid = () => {
  const { exp } = getAccessTokenDecoded()

  return Date.now() <= exp * 1000
}

export const isAuthenticated = () => {
  const sessionData = getSessionData()

  return sessionData.access_token && isTokenValid()
}

export const isAllowedByRole = (role: Authority[] = []) => {
  if (role.length === 0) return true
  const { authorities } = getAccessTokenDecoded()
  return role.some((el) => authorities?.includes(el))
}

export const logout = () => {
  localStorage.removeItem('authData')
  history.replace('/')
}
