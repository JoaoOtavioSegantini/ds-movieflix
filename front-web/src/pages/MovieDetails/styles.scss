.movie-details-container {
  width: 1170px;
  margin: 0 auto;
  margin-top: 35px;
  margin-bottom: 45px;
}

.movie-details {
  min-height: 374px;
}

.movie-main-image {
  padding: 24px 58px 30px 17px;
}

.card-description {
  border: 1px solid var(--border);
  width: 541px;
  font-size: 16px;
  line-height: 22px;
  text-align: justify;
  letter-spacing: -0.015em;
  color: var(--grey);
  padding: 13px 28px 10px 19px;
  margin-bottom: 15px;
}
.movie-main-details {
  margin-top: 24px;
}

.card-title-detail {
  font-weight: bold;
  font-size: 36px;
  line-height: 49px;
  letter-spacing: -0.015em;
  color: var(--pure-white);
}

.card-year-detail {
  font-weight: bold;
  font-size: 24px;
  line-height: 33px;
  letter-spacing: -0.015em;
  color: var(--primary-color);
}

.card-synopsis-detail {
  font-size: 18px;
  line-height: 25px;
  letter-spacing: -0.015em;
  color: var(--pure-white);
  margin-bottom: 9px;
}

.avaliation-container {
  border: 1px solid var(--grey);
  box-sizing: border-box;
  border-radius: 10px;
  width: 1131px;
  height: 74px;
  padding: 10px 25px 40px 16px;
  margin-top: 16px;
  margin-left: 17px;
  &::placeholder {
    font-size: 16px;
    line-height: 22px;
    letter-spacing: -0.015em;
    color: var(--grey);
  }
  &:disabled {
    cursor: not-allowed;
  }
}

.input-save-avaliation {
  min-height: 165px;
}
.save-avaliation {
  margin-top: 12px;
  margin-left: auto;
  margin-right: auto;
  display: flex;
  height: 50px;
  border: none;
  font-weight: bold;
  font-size: 16px;
  line-height: 22px;
  text-align: center;
  letter-spacing: -0.015em;
  color: var(--dark);
  margin-bottom: 13px;
  padding: auto 17.68px;
}

.invalid-save-review {
  cursor: not-allowed;
}

input:not(:checked) ~ p:not(.truncated) ~ label {
  display: none;
}

input.more-information {
  opacity: 0;
  position: absolute;
  pointer-events: none;

  &:checked + p {
    -webkit-line-clamp: unset;
  }
}

label {
  cursor: pointer;
  color: var(--primary-color);
  float: right;
  margin-top: -0.75em;
}

blockquote > p {
  padding-top: 2px;
  display: -webkit-box;
  -webkit-line-clamp: 7;
  -webkit-box-orient: vertical;
  overflow: hidden;
  margin-bottom: 22px;
}

@media (max-width: 768px) {
  .movie-details-container {
    width: 95vw;
  }
  .card-description {
    width: 80vw;
  }
  .movie-details {
    display: flex;
    flex-direction: column !important;
  }

  .movie-main-image {
    margin: 0 10vw;
  }

  .movie-main-details {
    margin: 0 10vw;
  }

  .input-save-avaliation {
    width: 95vw;
    margin: 0 0.5vw;
  }
  .avaliation-container {
    width: 90vw;
  }
}

@media (max-width: 576px) {
  .movie-details-container {
    width: 98%;
  }
  .movie-main-image > img {
    width: 130%;
    height: 120%;
    margin-left: -20px;
  }

  .card-title-detail {
    font-size: 1.8em;
  }
  .card-description {
    width: 100%;
  }
  .input-save-avaliation {
    margin: 0;
    width: 98vw;
    margin: 5vw 0.5vw !important;
  }
  blockquote > p {
    -webkit-line-clamp: 10;
  }

  .save-avaliation {
    width: 90vw !important;

  }

  .avaliation-container {
    margin-left: 4vw;
  }

}
