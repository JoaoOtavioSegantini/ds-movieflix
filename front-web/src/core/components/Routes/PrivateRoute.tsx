import { isAllowedByRole, isAuthenticated, Authority } from '@utils/auth'
import { Redirect, Route } from 'react-router'

type Props = {
  children: React.ReactNode
  path: string
  allowedRoutes?: Authority[]
  exact?: boolean
}

const PrivateRoute = ({ children, path, allowedRoutes, exact }: Props) => {
  return (
    <Route
      exact={exact}
      path={path}
      render={({ location }) => {
        if (!isAuthenticated()) {
          return (
            <Redirect
              to={{
                pathname: '/',
                state: { from: location }
              }}
            />
          )
        } else if (isAuthenticated() && !isAllowedByRole(allowedRoutes)) {
          return <Redirect exact to={{ pathname: '/movies' }} />
        }

        return children
      }}
    />
  )
}

export default PrivateRoute
