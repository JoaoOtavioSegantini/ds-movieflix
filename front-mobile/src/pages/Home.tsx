import { useNavigation } from "@react-navigation/native";
import React from "react";
import { Image, Text, TouchableOpacity, View } from "react-native";

import arrow from "../assets/arrow.svg";
import main from "../assets/main.png";
import { theme, text } from "../styles";

const Home: React.FC = () => {
  const navigation = useNavigation();

  return (
    <View style={theme.container}>
      <View style={theme.card}>
        <Image source={main} style={theme.mainImage} />
        <View style={theme.textContainer}>
          <Text style={text.bold}>Avalie filmes</Text>
          <Text style={text.regular}>
            Diga o que vocÃª achou do seu filme favorito
          </Text>
        </View>
        <TouchableOpacity
          activeOpacity={0.8}
          style={theme.primaryButton}
          onPress={() => navigation.navigate("Login" as never)}
        >
          <Text style={text.primaryText}>Fazer login</Text>
          <View style={theme.arrowContainer}>
            <Image source={arrow} style={{ width: 8, height: 15 }} />
          </View>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default Home;
