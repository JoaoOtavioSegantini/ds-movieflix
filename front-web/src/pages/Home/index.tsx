import './styles.scss'

import { ReactComponent as MainImage } from '@images/login.svg'
import { ReactComponent as ArrowIcon } from '@images/arrow.svg'

import { useHistory, useLocation } from 'react-router-dom'
import { useEffect, useState } from 'react'
import { useForm } from 'react-hook-form'

import { makeLogin } from '@utils/request'
import { saveSessionData } from '@utils/auth'
import Alert from '@components/Alert'
import ButtonLoader from '@components/ButtonLoader'
import IconShow from '@components/IconShow'
import Form from '@components/Form'
import { toast } from 'react-toastify'

export type FormState = {
  username: string
  password: string
}

type LocationState = {
  from: string
}

const Home = () => {
  const [values, setValues] = useState({
    password: '',
    showPassword: false
  })

  const {
    handleSubmit,
    register,
    formState: { errors }
  } = useForm<FormState>()

  const [hasError, setHasError] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const history = useHistory()
  const location = useLocation<LocationState>()

  const handleClickPassword = () => {
    setValues({ ...values, showPassword: !values.showPassword })
  }

  const handleMouseDownPassword = (event: React.MouseEvent) => {
    event.preventDefault()
  }

  const { from } = location.state || { from: { pathname: '/movies' } }

  const onSubmit = (data: FormState) => {
    setIsLoading(true)
    makeLogin(data)
      .then((response) => {
        setHasError(false)

        setIsLoading(false)
        saveSessionData(response.data)
        history.replace(from)
        setTimeout(() => {
          toast.success(`Bem vindo de volta ${response.data.userName}!`)
        }, 1500)
      })
      .catch((err) => {
        toast.error(err)
        setHasError(true)
        setIsLoading(false)
      })
  }

  return (
    <div className="home-container body-color">
      <div className="home-content">
        <div className="home-text col-6">
          <h1 className="text-title">Avalie Filmes</h1>
          <h1 className="text-subtitle">
            Diga o que vocÃª achou do seu filme favorito
          </h1>
          <MainImage className="main-image" data-testid="main-image" />
        </div>
        <form className="col-6 login-base" onSubmit={handleSubmit(onSubmit)}>
          <h1 className="login-title">LOGIN</h1>
          {hasError && <Alert onClick={() => setHasError(false)} />}
          <Form
            errors={errors}
            register={register}
            showText={values.showPassword}
          >
            <IconShow
              errors={!!errors.password}
              errorPasswordMessage={errors?.password?.message}
              showPassword={values.showPassword}
              handleClickPassword={handleClickPassword}
              handleMouseDownPassword={handleMouseDownPassword}
            />
          </Form>
          <div className="d-flex">
            <ButtonLoader
              isLoading={isLoading}
              disabled={!!errors.password || !!errors.username}
            />
            <div className="btn-icon-content" data-testid="arrowIcon">
              <ArrowIcon />
            </div>
          </div>
        </form>
      </div>
    </div>
  )
}

export default Home
