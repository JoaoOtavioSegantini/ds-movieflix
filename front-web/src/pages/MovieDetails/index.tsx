import './styles.scss'
//REACT_APP_WAIT_ON_TIMEOUT=600000
//REACT_APP_WAIT_ON_INTERVAL=600000

import UserAvaliation from '@components/UserAvaliation'
import { Movie } from '@type/Movie'
import { makePrivateRequest } from '@utils/request'
import { isAllowedByRole } from '@utils/auth'
import MovieCardDetailsLoader from '@components/Loaders/MovieCardDetailsLoader'
import SaveAvaliationLoader from '@components/Loaders/SaveAvaliationLoader'
import ReviewsLoader from '@components/Loaders/ReviewsLoader'
import { Modal } from '@components/InfoModal'

import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { toast } from 'react-toastify'
import ResizeObserver from 'resize-observer-polyfill'

type ParamsType = {
  movieId: string
}

const MovieDetails = () => {
  const { movieId } = useParams<ParamsType>()
  const [movie, setMovie] = useState<Movie>()
  const [load, setLoad] = useState(false)
  const [error, setError] = useState(false)
  const [moreInformation, setMoreInformation] = useState(false)

  useEffect(() => {
    setLoad(true)
    makePrivateRequest({ url: `/movies/${movieId}` })
      .then((response) => {
        setMovie(response.data)
      })
      .finally(() => {
        setLoad(false)
      })
      .catch((err) => toast.error(err.message))
  }, [movieId])

  useEffect(() => {
    const texts = document.querySelectorAll('p.description')

    const observer = new ResizeObserver((entries) => {
      for (const entry of entries) {
        entry.target.classList[
          entry.target.scrollHeight > entry.contentRect.height + 25
            ? 'add'
            : 'remove'
        ]('truncated')
      }
    })

    texts.forEach((text) => observer.observe(text))
  })

  const member = isAllowedByRole(['MEMBER'])

  const movieIdParsed = Number(movieId)

  const saveAvaliation = () => {
    if (!member) return
    const saveRev = document.querySelectorAll('input')[1]?.value
    if (!saveRev.match(/^(?!\s*$).+/)) return setError(true)

    makePrivateRequest({
      url: '/reviews',
      method: 'POST',
      data: { movieId: movieIdParsed, text: saveRev }
    })
      .then((userReview) => {
        toast.success(`Avaliação do filme ${movie?.title} enviada com sucesso!`)

        const movieReviews = movie?.reviews
        const reviewsUpdated = movieReviews!.concat(userReview.data)

        setMovie({ ...movie!, reviews: reviewsUpdated })

        const clearInput = document.querySelectorAll('input')[1]
        clearInput!.value = ''
      })
      .catch((err) => {
        toast.error(err.message)
        setError(false)
      })
  }

  const handleCloseModal = () => {
    setError(false)
  }

  return (
    <div className="movie-details-container">
      {load ? (
        <>
          <MovieCardDetailsLoader />
          <SaveAvaliationLoader />
          <ReviewsLoader />
        </>
      ) : (
        <>
          {error && <Modal close={handleCloseModal} />}
          <div className="movie-details d-flex card-base">
            <div className="movie-main-image">
              <img src={movie?.imgUrl} width={526} height={320} />
            </div>
            <div className="movie-main-details">
              <div className="card-title-detail" data-testid="movie-title">
                <span>{movie?.title}</span>
              </div>
              <div className="card-year-detail">
                <span>{movie?.year}</span>
              </div>
              <div className="card-synopsis-detail">
                <span>{movie?.subTitle}</span>
              </div>
              <div className="card-description border-radius-20">
                <blockquote>
                  <input
                    type="checkbox"
                    className="more-information"
                    id={`movie-${movie?.id}`}
                    data-testid="view-more"
                  />
                  <p className="description">{movie?.synopsis}</p>
                  <label
                    className="label-more"
                    htmlFor={`movie-${movie?.id}`}
                    onClick={() => setMoreInformation(!moreInformation)}
                  >
                    {moreInformation ? 'Resumir' : 'Ver tudo'}
                  </label>
                </blockquote>
              </div>
            </div>
          </div>
          <div className="input-save-avaliation card-base d-block mt-3">
            <input
              className="avaliation-container col-6"
              placeholder="Deixe sua avaliação aqui"
              data-testid="avaliation-save-input-main"
              disabled={!member}
              title={
                !member ? 'Somente membros podem salvar uma avaliação' : ''
              }
            />
            <div
              className={`btn btn-primary save-avaliation col-6 ${
                !member ? 'invalid-save-review' : ''
              }`}
              style={{ width: '334px' }}
              data-testid="review-onsave"
              onClick={saveAvaliation}
              title={
                !member ? 'Somente membros podem salvar uma avaliação' : ''
              }
            >
              SALVAR AVALIAÇÃO
            </div>
          </div>
          {movie?.reviews && movie.reviews.length > 0 && (
            <div className="user-avaliation card-base mt-3">
              {movie?.reviews.map((review) => (
                <UserAvaliation review={review} key={review.id} />
              ))}
            </div>
          )}
        </>
      )}
    </div>
  )
}

export default MovieDetails
