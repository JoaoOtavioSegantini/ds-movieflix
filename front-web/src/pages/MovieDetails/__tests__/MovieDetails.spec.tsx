import { render, screen, waitFor } from '@testing-library/react'
import { Route, Router, useParams, Switch } from 'react-router-dom'
import history from 'src/core/utils/history'
import MovieDetails from '..'
import { mock } from '@mock/detailsMock'
import { setupServer } from 'msw/node'
import { rest } from 'msw'
import userEvent from '@testing-library/user-event'
import SessionMock from '@mock/sessionMock'
import PrivateRoute from 'src/core/components/Routes/PrivateRoute'
import { act } from 'react-dom/test-utils'
import { ToastContainer } from 'react-toastify'
import Home from 'src/pages/Home'
import * as authMock from 'src/core/utils/auth'

const movieId = '1'

let success = true
let revSuccess = true

const review = [
  {
    id: 2,
    text: 'Achei o filme ótimo!!!',
    movieId: 1,
    user: {
      id: 1,
      email: 'joao@gmail.com',
      name: 'Joao',
      roles: [
        {
          id: 2,
          authority: 'MEMBER'
        }
      ]
    }
  }
]

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn()
}))

const server = setupServer(
  rest.get(`http://localhost:8080/movies/${movieId}`, (req, res, ctx) => {
    return !success
      ? res(ctx.status(500))
      : res(ctx.json(mock), ctx.status(200))
  }),
  rest.post('http://localhost:8080/reviews', (req, res, ctx) => {
    return !revSuccess
      ? res(ctx.status(404, 'simple text error message'))
      : res(ctx.status(200), ctx.json(review))
  })
)
beforeEach(() => {
  jest.clearAllMocks(),
    (useParams as jest.Mock).mockReturnValue({
      movieId: '1'
    })
})
beforeAll(() => {
  server.listen({ onUnhandledRequest: 'bypass' })
})
afterEach(() => {
  jest.clearAllMocks(), window.localStorage.clear(), server.resetHandlers()
})
afterAll(() => server.close())

describe('<MovieDetails />', () => {
  test('should render <MovieDetails /> component', async () => {
    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <MovieDetails />
        </Route>
      </Router>
    )

    const subTitle = await waitFor(() =>
      screen.getByText(
        'O vilão Loki reempreende seu papel como o Deus do Mal....'
      )
    )
    const year = screen.getByText(2021)
    const imgUrl = screen.getAllByRole('img')[0]
    const overview = screen.getByText(
      'Loki é uma série estadunidense criada por Michael Waldron para o Disney+, baseada no personagem de mesmo nome da Marvel Comics. É ambientada no Universo Cinematográfico Marvel (UCM), compartilhando continuidade com os filmes da franquia. A série se passa após os eventos do filme Vingadores: Ultimato (2019), no qual uma versão alternativa de Loki cria uma nova linha do tempo. Loki é produzida pelo Marvel Studios, com Waldron atuando como roteirista principal e Kate Herron dirigindo a primeira temporada.Tom Hiddleston repete seu papel como Loki da série de filmes, com Gugu Mbatha-Raw, Wunmi Mosaku, Eugene Cordero, Tara Strong, Owen Wilson, Sophia Di Martino, Sasha Lane, Jack Veal, DeObia Oparei, Richard E. Grant e Jonathan Majors também estrelando. Em setembro de 2018, o Marvel Studios estava desenvolvendo uma série de séries limitadas para o Disney+, centradas em personagens coadjuvantes dos filmes do UCM.Depois de roubar o Tesseract durante os eventos de Vingadores: Ultimato, uma versão alternativa de Loki é trazida para a misteriosa Autoridade de Variância Temporal (AVT), uma organização burocrática que existe fora do tempo e do espaço, e monitora a linha do tempo. Eles dão a Loki uma escolha: ser apagado da existência por ser uma "variante do tempo" ou ajudar a consertar a linha do tempo e impedir uma ameaça maior. Loki acaba preso em seu próprio thriller policial, viajando no tempo e alterando a história da humanidade.'
    )
    const userRev = screen.getByText('Alex Brown')
    const userText = screen.getByText(
      'Achei o filme ótimo, meus parabéns para o diretor!!!'
    )

    expect(history.location.pathname).toBe('/movies/1')
    expect(year).toBeInTheDocument()
    expect(year).toHaveStyle({ color: 'var(--primary-color)' })
    expect(screen.getByTestId('movie-title')).toBeInTheDocument()
    expect(screen.getByTestId('movie-title')).toHaveStyle({
      color: 'var(--pure-white)'
    })
    expect(imgUrl).toBeInTheDocument()
    expect(imgUrl).toHaveAttribute(
      'src',
      'https://www.themoviedb.org/t/p/w533_and_h300_bestv2/5fRAAO13URmteku8mb39V9YPJBb.jpg'
    )
    expect(subTitle).toBeInTheDocument()
    expect(overview).toBeInTheDocument()
    expect(userRev).toBeInTheDocument()
    expect(userText).toBeInTheDocument()
  })
  test('should render modal when user type empty review', async () => {
    new SessionMock().userAll()
    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <MovieDetails />
        </Route>
      </Router>
    )

    expect(screen.getAllByTitle('Loading...')).toHaveLength(3)

    await waitFor(async () => {
      const userRev = screen.getByTestId('avaliation-save-input-main')
      await userEvent.type(userRev, '')
    })
    const saveRev = screen.getByTestId('review-onsave')

    userEvent.click(saveRev)

    const buttonCloseModal = screen.getByRole('button')

    expect(document.querySelector('.overlay')).toBeInTheDocument()
    userEvent.click(buttonCloseModal)
    expect(buttonCloseModal.onclick?.call.length).toBe(1)
    expect(document.querySelector('.overlay')).not.toBeInTheDocument()
  })
  test('should render view more information label', async () => {
    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <MovieDetails />
        </Route>
      </Router>
    )
    await waitFor(() => {
      expect(screen.getByText('Ver tudo')).toBeInTheDocument()
      expect(screen.queryByText('Resumir')).not.toBeInTheDocument()
    })
    expect(document.querySelector('p')).not.toHaveClass('truncated')
    userEvent.click(screen.getByText('Ver tudo'))

    expect(screen.getByText('Resumir')).toBeInTheDocument()
    expect(screen.queryByText('Ver tudo')).not.toBeInTheDocument()
  })
  test(`should don't render disable save review when user is a member`, async () => {
    new SessionMock().userAll()
    history.push('/movies/1')
    const { container } = render(
      <Router history={history}>
        <PrivateRoute allowedRoutes={['MEMBER']} path="/movies/1">
          <MovieDetails />
        </PrivateRoute>
      </Router>
    )

    await waitFor(async () => {
      expect(
        screen.getByText(
          /O vilão Loki reempreende seu papel como o Deus do Mal..../i
        )
      ).toBeInTheDocument()
      await expect(
        screen.queryAllByTitle(/Somente membros podem salvar uma avaliação/i)
      ).toHaveLength(0)
    })
    await waitFor(() => expect(container.firstChild).toMatchSnapshot())
  })
  test('should render disable save review when use is not a member', async () => {
    history.push('/movies/1')
    const { container } = render(
      <Router history={history}>
        <Route path="/movies/1">
          <MovieDetails />
        </Route>
      </Router>
    )
    await waitFor(() => {
      expect(
        screen.getAllByTitle(/Somente membros podem salvar uma avaliação/i)
      ).toHaveLength(2)

      expect(screen.getByTestId('review-onsave')).toHaveClass(
        'invalid-save-review'
      )
    })
    expect(screen.getByTestId('avaliation-save-input-main')).toHaveAttribute(
      'disabled'
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  test('should send text review when user is a member', async () => {
    new SessionMock().userAll()

    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <ToastContainer />
          <MovieDetails />
        </Route>
      </Router>
    )
    await waitFor(() => {
      expect(
        screen.getByText(
          /O vilão Loki reempreende seu papel como o Deus do Mal..../i
        )
      ).toBeInTheDocument()
    })
    act(() => {
      const review = 'Achei o filme ótimo!!!'
      userEvent.type(
        screen.getByPlaceholderText(/deixe sua avaliação aqui/i),
        review
      )
      userEvent.click(screen.getByText(/salvar avaliação/i))
    })

    await waitFor(() =>
      expect(
        screen.getByText('Avaliação do filme Loki enviada com sucesso!')
      ).toBeInTheDocument()
    )
    expect(screen.getByRole('button', { name: /close/i })).toBeInTheDocument()
    expect(screen.getByText('Achei o filme ótimo!!!')).toBeInTheDocument()
    expect(screen.getByText('Joao')).toBeInTheDocument()
  })
  test('should show text error when API call fail', async () => {
    success = false
    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <ToastContainer />
          <MovieDetails />
        </Route>
      </Router>
    )
    await waitFor(() => {
      expect(
        screen.getByText('Request failed with status code 500')
      ).toBeInTheDocument()
    })

    act(() => {
      const review = 'Achei o filme ótimo!!!'
      userEvent.type(
        screen.getByPlaceholderText(/deixe sua avaliação aqui/i),
        review
      )
      userEvent.click(screen.getByText(/salvar avaliação/i))
    })
  })
  test('should show text error when reviews API call fail', async () => {
    success = true
    revSuccess = false
    new SessionMock().userAll()
    history.push('/movies/1')
    render(
      <Router history={history}>
        <Route path="/movies/1">
          <ToastContainer />
          <MovieDetails />
        </Route>
      </Router>
    )
    await waitFor(() => {
      expect(
        screen.getByText(
          /O vilão Loki reempreende seu papel como o Deus do Mal..../i
        )
      ).toBeInTheDocument()
    })

    act(() => {
      const review = 'Achei o filme ótimo!!!'
      userEvent.type(
        screen.getByPlaceholderText(/deixe sua avaliação aqui/i),
        review
      )
      userEvent.click(screen.getByText(/salvar avaliação/i))
    })

    await waitFor(() => {
      expect(
        screen.getByText('Request failed with status code 404')
      ).toBeInTheDocument()
    })
  })
  test('should redirect to home when user is unauthorized', async () => {
    jest.spyOn(authMock, 'logout')
    new SessionMock().userAll()

    const testErrorMessage = 'THIS IS A TEST FAILURE'

    server.use(
      rest.post(`http://localhost:8080/reviews`, async (req, res, ctx) => {
        return res(ctx.status(401), ctx.json({ message: testErrorMessage }))
      })
    )

    await waitFor(() =>
      render(
        <Router history={history}>
          <Switch>
            <Route path="/movies/1">
              <MovieDetails />
            </Route>
            <Route path="/" exact>
              <Home />
            </Route>
          </Switch>
        </Router>
      )
    )
    await waitFor(() => {
      expect(
        screen.getByText(
          /O vilão Loki reempreende seu papel como o Deus do Mal..../i
        )
      ).toBeInTheDocument()
    })

    act(() => {
      history.push('/movies/1')
      expect(window.location.pathname).toBe('/movies/1')
      const review = 'Achei o filme ótimo!!!'
      userEvent.type(screen.getByTestId('avaliation-save-input-main'), review)
      userEvent.click(screen.getByText(/salvar avaliação/i))
    })
    await waitFor(() => {
      expect(window.location.pathname).toBe('/')
      expect(authMock.logout).toHaveBeenCalled()
    })
  })
})
