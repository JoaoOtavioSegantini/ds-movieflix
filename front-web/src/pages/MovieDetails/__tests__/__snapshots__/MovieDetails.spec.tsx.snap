// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDetails /> should don't render disable save review when user is a member 1`] = `
<div
  class="movie-details-container"
>
  <div
    class="movie-details d-flex card-base"
  >
    <div
      class="movie-main-image"
    >
      <img
        height="320"
        src="https://www.themoviedb.org/t/p/w533_and_h300_bestv2/5fRAAO13URmteku8mb39V9YPJBb.jpg"
        width="526"
      />
    </div>
    <div
      class="movie-main-details"
    >
      <div
        class="card-title-detail"
        data-testid="movie-title"
      >
        <span>
          Loki
        </span>
      </div>
      <div
        class="card-year-detail"
      >
        <span>
          2021
        </span>
      </div>
      <div
        class="card-synopsis-detail"
      >
        <span>
          O vilão Loki reempreende seu papel como o Deus do Mal....
        </span>
      </div>
      <div
        class="card-description border-radius-20"
      >
        <blockquote>
          <input
            class="more-information"
            data-testid="view-more"
            id="movie-1"
            type="checkbox"
          />
          <p
            class="description"
          >
            Loki é uma série estadunidense criada por Michael Waldron para o Disney+, baseada no personagem de mesmo nome da Marvel Comics. É ambientada no Universo Cinematográfico Marvel (UCM), compartilhando continuidade com os filmes da franquia. A série se passa após os eventos do filme Vingadores: Ultimato (2019), no qual uma versão alternativa de Loki cria uma nova linha do tempo. Loki é produzida pelo Marvel Studios, com Waldron atuando como roteirista principal e Kate Herron dirigindo a primeira temporada.Tom Hiddleston repete seu papel como Loki da série de filmes, com Gugu Mbatha-Raw, Wunmi Mosaku, Eugene Cordero, Tara Strong, Owen Wilson, Sophia Di Martino, Sasha Lane, Jack Veal, DeObia Oparei, Richard E. Grant e Jonathan Majors também estrelando. Em setembro de 2018, o Marvel Studios estava desenvolvendo uma série de séries limitadas para o Disney+, centradas em personagens coadjuvantes dos filmes do UCM.Depois de roubar o Tesseract durante os eventos de Vingadores: Ultimato, uma versão alternativa de Loki é trazida para a misteriosa Autoridade de Variância Temporal (AVT), uma organização burocrática que existe fora do tempo e do espaço, e monitora a linha do tempo. Eles dão a Loki uma escolha: ser apagado da existência por ser uma "variante do tempo" ou ajudar a consertar a linha do tempo e impedir uma ameaça maior. Loki acaba preso em seu próprio thriller policial, viajando no tempo e alterando a história da humanidade.
          </p>
          <label
            class="label-more"
            for="movie-1"
          >
            Ver tudo
          </label>
        </blockquote>
      </div>
    </div>
  </div>
  <div
    class="input-save-avaliation card-base d-block mt-3"
  >
    <input
      class="avaliation-container col-6"
      data-testid="avaliation-save-input-main"
      placeholder="Deixe sua avaliação aqui"
      title=""
    />
    <div
      class="btn btn-primary save-avaliation col-6 "
      data-testid="review-onsave"
      style="width: 334px;"
      title=""
    >
      SALVAR AVALIAÇÃO
    </div>
  </div>
  <div
    class="user-avaliation card-base mt-3"
  >
    <div
      class="user-avaliation-container pt-3"
    >
      <div
        class="username"
      >
        <img
          alt="estrela amarela"
          src="icon-star.svg"
        />
        <span>
           
          Alex Brown
        </span>
      </div>
      <div
        class="comment border-radius-20"
      >
        <p>
          Achei o filme ótimo, meus parabéns para o diretor!!!
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<MovieDetails /> should render disable save review when use is not a member 1`] = `
<div
  class="movie-details-container"
>
  <div
    class="movie-details d-flex card-base"
  >
    <div
      class="movie-main-image"
    >
      <img
        height="320"
        src="https://www.themoviedb.org/t/p/w533_and_h300_bestv2/5fRAAO13URmteku8mb39V9YPJBb.jpg"
        width="526"
      />
    </div>
    <div
      class="movie-main-details"
    >
      <div
        class="card-title-detail"
        data-testid="movie-title"
      >
        <span>
          Loki
        </span>
      </div>
      <div
        class="card-year-detail"
      >
        <span>
          2021
        </span>
      </div>
      <div
        class="card-synopsis-detail"
      >
        <span>
          O vilão Loki reempreende seu papel como o Deus do Mal....
        </span>
      </div>
      <div
        class="card-description border-radius-20"
      >
        <blockquote>
          <input
            class="more-information"
            data-testid="view-more"
            id="movie-1"
            type="checkbox"
          />
          <p
            class="description"
          >
            Loki é uma série estadunidense criada por Michael Waldron para o Disney+, baseada no personagem de mesmo nome da Marvel Comics. É ambientada no Universo Cinematográfico Marvel (UCM), compartilhando continuidade com os filmes da franquia. A série se passa após os eventos do filme Vingadores: Ultimato (2019), no qual uma versão alternativa de Loki cria uma nova linha do tempo. Loki é produzida pelo Marvel Studios, com Waldron atuando como roteirista principal e Kate Herron dirigindo a primeira temporada.Tom Hiddleston repete seu papel como Loki da série de filmes, com Gugu Mbatha-Raw, Wunmi Mosaku, Eugene Cordero, Tara Strong, Owen Wilson, Sophia Di Martino, Sasha Lane, Jack Veal, DeObia Oparei, Richard E. Grant e Jonathan Majors também estrelando. Em setembro de 2018, o Marvel Studios estava desenvolvendo uma série de séries limitadas para o Disney+, centradas em personagens coadjuvantes dos filmes do UCM.Depois de roubar o Tesseract durante os eventos de Vingadores: Ultimato, uma versão alternativa de Loki é trazida para a misteriosa Autoridade de Variância Temporal (AVT), uma organização burocrática que existe fora do tempo e do espaço, e monitora a linha do tempo. Eles dão a Loki uma escolha: ser apagado da existência por ser uma "variante do tempo" ou ajudar a consertar a linha do tempo e impedir uma ameaça maior. Loki acaba preso em seu próprio thriller policial, viajando no tempo e alterando a história da humanidade.
          </p>
          <label
            class="label-more"
            for="movie-1"
          >
            Ver tudo
          </label>
        </blockquote>
      </div>
    </div>
  </div>
  <div
    class="input-save-avaliation card-base d-block mt-3"
  >
    <input
      class="avaliation-container col-6"
      data-testid="avaliation-save-input-main"
      disabled=""
      placeholder="Deixe sua avaliação aqui"
      title="Somente membros podem salvar uma avaliação"
    />
    <div
      class="btn btn-primary save-avaliation col-6 invalid-save-review"
      data-testid="review-onsave"
      style="width: 334px;"
      title="Somente membros podem salvar uma avaliação"
    >
      SALVAR AVALIAÇÃO
    </div>
  </div>
  <div
    class="user-avaliation card-base mt-3"
  >
    <div
      class="user-avaliation-container pt-3"
    >
      <div
        class="username"
      >
        <img
          alt="estrela amarela"
          src="icon-star.svg"
        />
        <span>
           
          Alex Brown
        </span>
      </div>
      <div
        class="comment border-radius-20"
      >
        <p>
          Achei o filme ótimo, meus parabéns para o diretor!!!
        </p>
      </div>
    </div>
  </div>
</div>
`;
